### ISSUE TITLE
wasm plugin memory leak in Traefik v3

### ISSUE THREAD

**Issue by ttys3** [traefik/traefik#11119]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you do?

add a simple wasm traefik middleware plugin to a router. 
run ab benchmark to make some request
you will see memory increase
stop the benchmark and wait
you will see the memory keep leaks, no matter how long you waited

the more concurrent requests make, the more memory leak.

@juliens  @zetaab  could you plz take a look at this?

here's the traefik memory usage grafana graph from our real server,  after enabling wasm based plugin:

![image](https://github.com/user-attachments/assets/21f2a491-0066-4e63-aab5-2f3125cb7280)

----------------------------------------------

steps to reproduce

go version 1.23.1
under Linux amd64 machine

```shell
git clone https://github.com/ttys3/traefik-wasm-plugin-memory-leak-demo.git

cd traefik-wasm-plugin-memory-leak-demo

pushd plugins-local/src/github.com/traefik/plugindemowasm/
# build wasm plugin
./build.sh

# you will see generated plugin.wasm now

# now we go back to traefik-wasm-plugin-memory-leak-demo dir
popd

# download newest traefik
curl -LO https://github.com/traefik/traefik/releases/download/v3.1.4/traefik_v3.1.4_linux_amd64.tar.gz
tar xvf traefik_v3.1.4_linux_amd64.tar.gz

# run the demo backend service, it is a simple http static server
./run-downstream-service.sh

# run traefik
./run.sh
```

now install ps_mem if you do not have it

```shell
sudo pip install ps_mem
```

now we open a new terminal to watch the memory

```shell
./watch-mem.sh
```

we 'll see that the init memory usage is: 

![image](https://github.com/user-attachments/assets/0abfede5-c5ab-457e-8c12-fea266431139)

go we run `./bench.sh` in the repo to make some requests (you should install `ab` - the Apache HTTP server benchmarking tool, if you os does not installed it, or other bench tools also OK)

> adjust params of `ab` if you system does not have too much free memory, in case OOM

```shell
./bench.sh

Benchmarking localhost (be patient)
Completed 100 requests
Completed 200 requests
Completed 300 requests
Completed 400 requests
Completed 500 requests
Completed 600 requests
Completed 700 requests
Completed 800 requests
Completed 900 requests
Completed 1000 requests
Finished 1000 requests


Server Software:        ECAcc
Server Hostname:        localhost
Server Port:            6688

Document Path:          /
Document Length:        1256 bytes

Concurrency Level:      100
Time taken for tests:   6.276 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      1632250 bytes
HTML transferred:       1256000 bytes
Requests per second:    159.33 [#/sec] (mean)
Time per request:       627.634 [ms] (mean)
Time per request:       6.276 [ms] (mean, across all concurrent requests)
Transfer rate:          253.97 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.4      0       2
Processing:   184  537 905.7    200    4695
Waiting:      184  536 905.4    200    4695
Total:        184  537 906.0    200    4696

Percentage of the requests served within a certain time (ms)
  50%    200
  66%    213
  75%    248
  80%    321
  90%   2040
  95%   3256
  98%   3400
  99%   3542
 100%   4696 (longest request)
```

### What did you see instead?

traefik memory keeps growing and never free

![image](https://github.com/user-attachments/assets/0ea48e7f-232c-413b-8c38-618424bcc1d0)

![image](https://github.com/user-attachments/assets/f46e6c6d-95b6-4ff7-b62d-22984be2febd)

![image](https://github.com/user-attachments/assets/92d50069-bba4-45aa-91a1-7f39ab7cbd74)

now we run 

```
go tool pprof http://127.0.0.1:8080/debug/pprof/heap
```
type `web` and enter, will see the heap graph:

![image](https://github.com/user-attachments/assets/5654eaee-764e-4de4-8a8a-58a760e387c6)

and here is the pprof file:

[pprof.traefik.alloc_objects.alloc_space.inuse_objects.inuse_space.009.pb.gz](https://github.com/user-attachments/files/17116655/pprof.traefik.alloc_objects.alloc_space.inuse_objects.inuse_space.009.pb.gz)

and also the allocs

![image](https://github.com/user-attachments/assets/fca8f4aa-c2ed-4fcb-b421-d6f85c8c3539)

allocs pprof file

[pprof.traefik.alloc_objects.alloc_space.inuse_objects.inuse_space.010.pb.gz](https://github.com/user-attachments/files/17116665/pprof.traefik.alloc_objects.alloc_space.inuse_objects.inuse_space.010.pb.gz)

note:

> the plugindemowasm plugin here is intended to demostrate the memory leak issue.

> not for real usage.

### What version of Traefik are you using?

```
 ./traefik version                    
Version:      3.1.4
Codename:     comte
Go version:   go1.23.1
Built:        2024-09-19T13:47:17Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration?

```yaml
# Static configuration

log:
  level: TRACE

global:
  checkNewVersion: false
  sendAnonymousUsage: false

api:
  dashboard: true
  insecure: true
  debug: true

metrics:
  prometheus: {}

entryPoints:
  web:
   address: ":6688"

providers:
  # https://doc.traefik.io/traefik/providers/file/
  file:
    watch: true
    filename: "dyn.yaml"

experimental:
  localPlugins:
    demowasm:
      moduleName: github.com/traefik/plugindemowasm

```

Add more configuration information here.

`dyn.yaml`

```yaml
# Dynamic configuration

http:
  routers:
    my-router:
      rule: Host(`localhost`)
      service: service-foo
      entryPoints:
        - web
      middlewares:
        - demowasm

  services:
   service-foo:
      loadBalancer:
        passHostHeader: false
        servers:
          - url: http://www.example.com/

  middlewares:
    demowasm:
      plugin:
        demowasm:
          headers:
            X-Demo: testxxxxx
            X-URL: 'http://example.localhost/'

```

### test 2

if we disable the wasm based plugin in `dyn.yaml` :

```diff
diff --git a/dyn.yaml b/dyn.yaml
index a1ce928..f71cdec 100755
--- a/dyn.yaml
+++ b/dyn.yaml
@@ -8,7 +8,7 @@ http:
       entryPoints:
         - web
       middlewares:
-        - demowasm
+        # - demowasm
 
   services:
    service-foo:
```

and run `./bench.sh` again, only cost  about100MB memory:

![image](https://github.com/user-attachments/assets/8f9fed5a-8b36-48d3-b5d9-59532578ed2a)

----------------------------

### maybe related

https://github.com/jcchavezs/coraza-http-wasm-traefik/issues/9

https://community.traefik.io/t/traefik-v3-coraza-waf-high-resource-usage/22814/8

https://github.com/tetratelabs/proxy-wasm-go-sdk/issues/450

also tried https://github.com/wasilibs/nottinygc, but no difference

**Comment by ttys3** [comment]
apply dynamic config (no http requests):

go pprotf heap

![pprof001](https://github.com/user-attachments/assets/ff07eb1d-b51f-4e24-898e-d41bbcae250b)

**Comment by juliens** [comment]
Thank you for the clarification. If I understand correctly, you're saying that the temporary memory usage increases but everything returns to normal after the GC runs.

Just to confirm, does this mean that there are no memory leaks because all the memory is released during GC, but the application still consumes a significant amount of memory during active use that we may can optimize because of the 1377 times the wasm plugin is created?

**Comment by ttys3** [comment]
@juliens I already find the solution, is it OK to make a PR ?

the memory leak code seems in github.com/traefik/http-wasm-host-go

go handles sync.Pool GC fine, but the pool object's inner type seems too complex and GC can not its inner object.

we need SetFinalizer to free that.

**Comment by juliens** [comment]
Good catch and for sure you can make a pull request.

Please make the PR on both the official repo ( https://github.com/http-wasm/http-wasm-host-go ) and the Traefik fork, the goal is to be able to go back to the official repo when it will possible.

Thanks again for your help!

**Comment by ttys3** [comment]
@juliens  here's the PR https://github.com/traefik/http-wasm-host-go/pull/2

**Comment by ttys3** [comment]
> Thank you for the clarification. If I understand correctly, you're saying that the temporary memory usage increases but everything returns to normal after the GC runs.

no. leak = memory still can not free after GC runs 

because everytime a traefik wasm plugin instance created (dynamic config loading or init traefik start can both trigger this), it only create one guest module usage from sync.Pool, so if the service route depends on the wasm middleware  still exists and no http request to it, it will keep the init memory usage. this "temporary memory" (in sync.Pool) will last for ever and here.

after with my PR, this will be no problem because the early inited first guest module, will all get GC ed corretly.
 
> Just to confirm, does this mean that there are no memory leaks because all the memory is released during GC, but the application still consumes a significant amount of memory during active use that we may can optimize because of the 1377 times the wasm plugin is created?

no. the application still consumes a significant amount of memory during NO single one http request.
that is what I said memory leak.

1377 times the wasm plugin is created? (see my answer to the first quote)

**Comment by ttys3** [comment]
prev I was wrong, the plugin init memory leak still exists:

no http request + make changes to `dyn.yaml` with 51 service has the wasm plugin middleware enabled
for 100 times

the memory leak keeps, the problem exists, because, new wasm host and guest created each time, but the memory not GC ed.  we have call `nethttp.middleware.Close()` (which will call `handler.middleware.Close()` -> wazero.Runtime.Close() ) to free the memory

![image](https://github.com/user-attachments/assets/c033f4ac-d0ab-43fa-9b64-6f193a697551)

**Comment by jcchavezs** [comment]
> everytime the WASM guest middleware handler method executes, this list still gets allocated in memory. Is there any way to support this on the host level so that all guest modules can share only one read-only list?

every guest instance will load rules in memory and I don't think they can be shared ATM cc @evacchi

**Comment by jcchavezs** [comment]
After looking at the code in Traefik I don't see it closing the middleware which must be called to clean up guest modules https://github.com/http-wasm/http-wasm-host-go/blob/9bfc11c198fd8c68f30659948a92580340cdaec5/handler/middleware.go#L190-L193. @juliens is there any syntax to call closer on plugins? I believe all plugins or middleware should be closed after replacing them.

**Comment by ttys3** [comment]
> After looking at the code in Traefik I don't see it closing the middleware which must be called to clean up guest modules [http-wasm/http-wasm-host-go@`9bfc11c`/handler/middleware.go#L190-L193](https://github.com/http-wasm/http-wasm-host-go/blob/9bfc11c198fd8c68f30659948a92580340cdaec5/handler/middleware.go#L190-L193). @juliens is there any syntax to call closer on plugins? I believe all plugins or middleware should be closed after replacing them.

yes, if traefik could call the `Close()` method on the middleware, then, the second `runtime.SetFinalizer` on my PR is unnecessary.

**Comment by juliens** [comment]
In fact, I think we can not call the close in Traefik without using SetFinalizer, because when traefik apply a new dynamic config, it switch the "global http.Handler" ( with the new middlewares in it ) but the previous "global http.Handler" can still be used by on going request, then the middleware can be Close only when the previous "global http.Handler" is GC when no more requests are using it.

**Comment by juliens** [comment]
https://github.com/traefik/traefik/compare/v3.1...juliens:traefik:fix-wasm-leak

@ttys3 with this, do we still need your other `SetFinalizer` on the pool because, I think that the close of the middleware, close the runtime which will close all the guest modules.

WDYT?

**Comment by jcchavezs** [comment]
la to the party https://github.com/traefik/traefik/pull/11150

**Comment by jcchavezs** [comment]
> I think that the close of the middleware, close the runtime which will close all the guest modules.

exactly

**Comment by ttys3** [comment]
@juliens @jcchavezs  here goes https://github.com/traefik/traefik/pull/11151

**Comment by ttys3** [comment]
> [v3.1...juliens:traefik:fix-wasm-leak](https://github.com/traefik/traefik/compare/v3.1...juliens:traefik:fix-wasm-leak)
> 
> @ttys3 with this, do we still need your other `SetFinalizer` on the pool because, I think that the close of the middleware, close the runtime which will close all the guest modules.
> 
> WDYT?

no, this is not correct. see my comment here: https://github.com/traefik/traefik/issues/11119#issuecomment-2388444040

**Comment by ttys3** [comment]
confirmed the leak issue resolved in traefik 3.1.6


after GC:

![Image](https://github.com/user-attachments/assets/56567a24-7199-4b30-bfce-2f0c09181461)

**Comment by evacchi** [comment]
hey all, I've followed this, I am happy this was not due to an upstream bug. Thanks!
